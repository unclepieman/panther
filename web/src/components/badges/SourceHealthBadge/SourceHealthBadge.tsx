/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { Badge, Flex, Icon, Tooltip, Text, Box } from 'pouncejs';
import { IntegrationItemHealthStatus } from 'Generated/schema';
import { slugify } from 'Helpers/utils';

interface SourceHealthBadgeProps {
  healthMetrics: IntegrationItemHealthStatus[];
}

const SourceHealthBadge: React.FC<SourceHealthBadgeProps> = ({ healthMetrics }) => {
  const isHealthy = healthMetrics.every(healthMetric => Boolean(healthMetric.healthy));

  const tooltipContent = (
    <Flex direction="column" spacing={1}>
      {healthMetrics.map(healthMetric => {
        const id = slugify(healthMetric.message);
        return (
          <Flex spacing={2} key={id}>
            <Icon
              mt="-2px" // we need that due to some alignment needs with `rawErrorMessage`
              aria-labelledby={id}
              size="medium"
              type={healthMetric.healthy ? 'check' : 'remove'}
              color={healthMetric.healthy ? 'green-400' : 'red-300'}
              aria-label={healthMetric.healthy ? 'Passing' : 'Failing'}
            />
            <Box>
              <Text id={id} aria-describedby={`${id}-description`}>
                {healthMetric.message}
              </Text>
              {!!healthMetric.rawErrorMessage && (
                <Text
                  my={1}
                  fontSize="x-small"
                  color="red-200"
                  id={`${id}-description`}
                  maxWidth="fit-content"
                >
                  {healthMetric.rawErrorMessage}
                </Text>
              )}
            </Box>
          </Flex>
        );
      })}
    </Flex>
  );

  const icon = isHealthy ? (
    <Badge color="green-400">HEALTHY</Badge>
  ) : (
    <Badge color="red-300">UNHEALTHY</Badge>
  );

  return <Tooltip content={tooltipContent}>{icon}</Tooltip>;
};

export default React.memo(SourceHealthBadge);
