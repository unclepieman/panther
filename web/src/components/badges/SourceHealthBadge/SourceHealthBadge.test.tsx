/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { buildIntegrationItemHealthStatus, render, fireEvent } from 'test-utils';
import SourceHealthBadge from './index';

describe('SourceHealthBadge', () => {
  it('matches original snapshot', () => {
    const { container } = render(
      <SourceHealthBadge healthMetrics={[buildIntegrationItemHealthStatus()]} />
    );

    expect(container).toMatchSnapshot();
  });

  it('correctly displays "HEALTHY" message', () => {
    const healthMetrics = [buildIntegrationItemHealthStatus({ healthy: true })];
    const { getByText } = render(<SourceHealthBadge healthMetrics={healthMetrics} />);

    expect(getByText('HEALTHY')).toBeInTheDocument();
  });

  it('correctly displays "UNHEALTHY" message', () => {
    const healthMetrics = [buildIntegrationItemHealthStatus({ healthy: false })];
    const { getByText } = render(<SourceHealthBadge healthMetrics={healthMetrics} />);

    expect(getByText('UNHEALTHY')).toBeInTheDocument();
  });

  it('correctly displays passing & failing health checks', async () => {
    const healthMetrics = [
      buildIntegrationItemHealthStatus({ healthy: true, message: 'Healthy Message' }),
      buildIntegrationItemHealthStatus({ healthy: false, message: 'Unhealthy Message' }),
    ];
    const { getByText, findByText, findByAriaLabel } = render(
      <SourceHealthBadge healthMetrics={healthMetrics} />
    );

    fireEvent.mouseOver(getByText('UNHEALTHY'));

    expect(await findByAriaLabel('Passing')).toBeInTheDocument();
    expect(await findByText('Healthy Message')).toBeInTheDocument();

    expect(await findByAriaLabel('Failing')).toBeInTheDocument();
    expect(await findByText('Healthy Message')).toBeInTheDocument();
  });

  it('shows raw error message for failing health checks', async () => {
    const healthMetrics = [
      buildIntegrationItemHealthStatus({
        healthy: false,
        message: 'Unhealthy Message',
        rawErrorMessage: 'Raw Error',
      }),
    ];

    const { getByText, findByText } = render(<SourceHealthBadge healthMetrics={healthMetrics} />);

    fireEvent.mouseOver(getByText('UNHEALTHY'));
    expect(await findByText('Raw Error')).toBeInTheDocument();
  });
});
