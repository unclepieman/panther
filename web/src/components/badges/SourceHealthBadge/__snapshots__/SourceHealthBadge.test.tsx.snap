// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceHealthBadge matches original snapshot 1`] = `
.panther-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: default;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  min-width: 85px;
  text-align: center;
  font-weight: 700;
  border-radius: 40px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  font-size: 0.75rem;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 16px;
  padding-right: 16px;
  background-color: #d64242;
}

<div>
  <header
    id="main-header"
  />
  <div
    aria-atomic="true"
    class="panther-0"
    data-reach-tooltip-trigger=""
    role="status"
  >
    UNHEALTHY
  </div>
  <footer
    id="footer"
  />
</div>
`;
